PR: https://github.com/odoo/odoo/pull/

From: 271c6223cb10230dc8a4d0b257d03f319a1fe790
From: Kamlesh Pathekar
Date: 2021-12-16 08:10:08

Structural Changes: 8
Total Changes: 10

[IMP] project: add analytic tags on project

-if an analytic tags are inserted on the project, it will be added to the timesheet /invoice
 analytic tag.
-In this commit, analytic tags field is added on project in order to gain in
 granularity.Those tags are passed to the invoice lines generated from the
 project's SO lines.
-set the analytic account and analytic tag of the project by default on its tasks.
-added string 'Analytic Tags' analytic_tag_ids field on task.

closes #78973
task-2667754

================================= pseudo patch: =================================

--- a/addons/project/models/analytic_account_tag.py
+++ b/addons/project/models/analytic_account_tag.py
@@ -7,3 +7,4 @@ class AccountAnalyticTag(models.Model):
     _inherit = 'account.analytic.tag'
 
     task_ids = fields.Many2many('project.task', string='Tasks')
+    project_ids = fields.Many2many('project.project', string='Projects')

--- a/addons/project/models/project.py
+++ b/addons/project/models/project.py
@@ -225,6 +225,7 @@ class Project(models.Model):
         help="Analytic account to which this project is linked for financial management. "
              "Use an analytic account to record cost and revenue on your project.")
     analytic_account_balance = fields.Monetary(related="analytic_account_id.balance")
+    analytic_tag_ids = fields.Many2many('account.analytic.tag', string='Analytic Tags')
 
     favorite_user_ids = fields.Many2many(
         'res.users', 'project_favorite_user_rel', 'project_id', 'user_id',
@@ -1111,7 +1112,7 @@ class Task(models.Model):
              "Use an analytic account to record cost and revenue on your task. "
              "If empty, the analytic account of the project will be used.")
     project_analytic_account_id = fields.Many2one('account.analytic.account', string='Project Analytic Account', related='project_id.analytic_account_id')
-    analytic_tag_ids = fields.Many2many('account.analytic.tag',
+    analytic_tag_ids = fields.Many2many('account.analytic.tag', string="Analytic Tags",
         domain="['|', ('company_id', '=', False), ('company_id', '=', company_id)]", check_company=True)
 
     @property
@@ -1565,6 +1566,12 @@ class Task(models.Model):
                 )
                 if partner_id:
                     vals['partner_id'] = partner_id
+        if vals.get('project_id'):
+            project = self.env['project.project'].browse(vals.get('project_id'))
+            if project.analytic_account_id:
+                vals['analytic_account_id'] = project.analytic_account_id.id
+            if project.analytic_tag_ids:
+                vals['analytic_tag_ids'] = [Command.set(project.analytic_tag_ids.ids)]
 
         return vals
 
